opal/include/opal/sys/atomic_stdc.h

```diff
@@ -215,13 +230,21 @@ typedef atomic_flag opal_atomic_lock_t;
  */
 static inline void opal_atomic_lock_init (opal_atomic_lock_t *lock, bool value)
 {
+#ifdef OPAL_HAVE_CLANG_BUILTIN_ATOMIC_C11_FUNC
```

Comment:
Can you say why we need to distinguish between Clang and the rest of the pack? According to https://en.cppreference.com/w/c/atomic/atomic_flag_clear `atomic_flag_clear` is supposed to take `volatile atomic_flag*` as argument (not `volatile void*`, which is now passed for the rest).

Reply:
It seems that clang uses it's own builtin-c11 atomic functions:
```
[awlauria@f8n02 clang_build]$ cat clang_test.c 
#include <stdatomic.h>

int main() {
 void *lock;
 atomic_flag_clear ((volatile void *) lock);
}
[awlauria@f8n02 clang_build]$ clang clang_test.c 
clang_test.c:5:2: error: member reference base type 'volatile void' is not a structure or union
 atomic_flag_clear ((volatile void *) lock);
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib64/clang/9.0.1/include/stdatomic.h:167:63: note: expanded from macro 'atomic_flag_clear'
#define atomic_flag_clear(object) __c11_atomic_store(&(object)->_Value, 0, __ATOMIC_SEQ_CST)
                                                      ~~~~~~~~^ ~~~~~~
1 error generated.
```

Checking with -E:

```
int main() {
 void *lock;
 __c11_atomic_store(&((volatile void *) lock)->_Value, 0, 5);
}
```
Dropping the volatile and just using a void * is fine.

Reply:
I think gcc/others will do an implicit cast and not warn as long as the pointer is volatile.

Reply:
Hmm, clang doesn't seem to like it either:

```
[awlauria@f8n02 clang_build]$ clang clang_test.c 
clang_test.c:5:2: error: member reference base type 'volatile atomic_flag *' (aka 'volatile struct atomic_flag *') is not a structure or union
 atomic_flag_clear (lock);
 ^~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib64/clang/9.0.1/include/stdatomic.h:167:63: note: expanded from macro 'atomic_flag_clear'
#define atomic_flag_clear(object) __c11_atomic_store(&(object)->_Value, 0, __ATOMIC_SEQ_CST)
                                                      ~~~~~~~~^ ~~~~~~
1 error generated.
```

gcc works though.

Reply:
I can make that change today - will test it across the different compilers.

Reply:
Why pass `void*` in the first place? Why not cast to `volatile atomic_lock*`?

Reply:
It looks like pgi replaces `atomic_clear(volatile atomic_lock *)` with  `__atomic_clear(volatile void * ptr, int memorder)`, and doesn't like the cast to volatile atomic_lock:

pgcc -E clang_test.c gives me:
```
int main() {
 volatile atomic_flag*  lock;
 __atomic_clear ((lock), 5);
}
```
I can't seem to make it happy without the cast to a volatile void *.
Maybe this should be a more of a PGI specific check.

Reply:
This seems like pleasing a broken compiler then. The definition of `atomic_flag_clear` is clear: the argument is a `volatile atomic_flag*`. If the PGI compiler trips on that its their problem, not ours :)

Reply:
I think (not positive) pgi is using these atomic built-ins:
https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html

which defines it as `Built-in Function: void __atomic_clear (bool *ptr, int memorder)`
Not sure if it's technically doing the 'wrong thing'.

Even if so - it might be worth a work-around, since this is throwing a few thousand warnings as it's propagated in many places.

Reply:
If others agree that we want to workaround the PGI compiler here, we should report it to PGI and have a special case for PGI, treating all others the "right way" (instead of having a workaround for the one compiler that trips over our workaround)